pkg/syncer/status/status_process_test.go

```diff
@@ -539,32 +591,81 @@ func TestStatusSyncerProcess(t *testing.T) {
+				func(o *metav1.ListOptions) {
+					o.LabelSelector = workloadv1alpha1.InternalDownstreamClusterLabel + "=" + syncTargetKey
+				},
+				&mockedGVRSource{},
+				cache.Indexers{
+					indexers.ByNamespaceLocatorIndexName: indexers.IndexByNamespaceLocator,
+				},
+			)
 			require.NoError(t, err)
 
-			toInformers.ForResource(tc.gvr).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+			var cacheSyncsForAlwaysRequiredGVRs []cache.InformerSynced
+			for _, alwaysRequired := range []string{"secrets", "namespaces"} {
+				if informer, err := ddsifForUpstreamSyncer.ForResource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: alwaysRequired}); err != nil {
+					require.NoError(t, err)
+				} else {
+					cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
+				}
+				if informer, err := ddsifForDownstream.ForResource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: alwaysRequired}); err != nil {
+					require.NoError(t, err)
+				} else {
+					cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
+				}
+			}
 
-			fromInformers.Start(ctx.Done())
-			toInformers.Start(ctx.Done())
+			setupServersideApplyPatchReactor(toClusterClient)
+			fromClientResourceWatcherStarted := setupWatchReactor(t, tc.gvr.Resource, fromClient)
+			toClientResourceWatcherStarted := setupClusterWatchReactor(t, tc.gvr.Resource, toClusterClient)
+
+			controller, err := NewStatusSyncer(logger, kcpLogicalCluster, tc.syncTargetName, syncTargetKey, tc.advancedSchedulingEnabled, toClusterClient, fromClient, ddsifForUpstreamSyncer, ddsifForDownstream, tc.syncTargetUID)
+			require.NoError(t, err)
 
-			fromInformers.WaitForCacheSync(ctx.Done())
-			toInformers.WaitForCacheSync(ctx.Done())
+			gvrsUpdated := make(chan struct{})
+			upstreamSyncerDDSIFUpdated := ddsifForDownstream.Subscribe("upstreamSyncer")
+			downstreamDDSIFUpdated := ddsifForDownstream.Subscribe("downstream")
+			go func() {
+				<-upstreamSyncerDDSIFUpdated
+				t.Logf("%s: upstream ddsif synced", t.Name())
+				<-downstreamDDSIFUpdated
+				t.Logf("%s: downstream ddsif synced", t.Name())
+
+				_, unsynced := ddsifForUpstreamSyncer.Listers()
+				for _, gvr := range unsynced {
+					informer, _ := ddsifForUpstreamSyncer.ForResource(gvr)
+					cache.WaitForCacheSync(ctx.Done(), informer.Informer().HasSynced)
+					t.Logf("%s: upstream ddsif informer synced for gvr %s", t.Name(), gvr.String())
+				}
+				_, unsynced = ddsifForDownstream.Listers()
+				for _, gvr := range unsynced {
+					informer, _ := ddsifForDownstream.ForResource(gvr)
+					cache.WaitForCacheSync(ctx.Done(), informer.Informer().HasSynced)
+					t.Logf("%s: downstream ddsif informer synced for gvr %s", t.Name(), gvr.String())
+				}
+				t.Logf("%s: gvrs updated", t.Name())
+				gvrsUpdated <- struct{}{}
+			}()
```

Comment:
is there no simpler way to test the controller? This complex synchronization code is a warning sign.

Reply:
It is related to the fact that we still create the StatusSyncer through the `NewStatusSyncer()` method, instead of the directly creating the StatusSyncer struct. So we need to pass real ddsif instances (similar to how previously we were passing the real informer factories), because these tests are still using fake client-go clients with reactors.

As mentioned in [this comment answer](https://github.com/kcp-dev/kcp/pull/2452#discussion_r1063557352), I had thought about switching the a fake-client-go-free, reactor-free test implementation, with only fake lister implementations.
I'd prefer keeping that for a distinct PR though.